# **Notebook entregable**

Incluye:

- Exploración y limpieza de datos.
- Tratamiento de variables financieras y textual
- Ingeniería de características
- Clasificación de gastos
- Análisis financiero
- Entrenamiento y evaluación de modelos
- Métricas de desempeño
- Serialización de modelos (.pkl o .joblib)

## **Exploración de datos**

### **Carga de librerías y dataset**

In [24]:
import pandas as pd
import unicodedata

pd.set_option("display.max_columns", None)
pd.set_option("display.max_colwidth", 100)

ruta = "dataset_finance_ai.csv"

df = pd.read_csv(ruta)

print("Dimensiones:", df.shape)
df.head()

Dimensiones: (4676, 13)


,user_id,nombre,arquetipo_perfil,periodo_inicio,periodo_fin,fecha,descripcion,monto,tipo_movimiento,categoria,grupo,tipo_pago,tasa_de_interes_de_la_tarjeta
0,110504306,Valentina Romero,Usuario con ingresos estacionales,2026-07-01,2026-07-31,2026-07-01,cosas para cocinar,164.68,Consumo,Alimentación,Esencial,Efectivo,NaN
1,110504306,Valentina Romero,Usuario con ingresos estacionales,2026-07-01,2026-07-31,2026-07-01,apoyo a un familiar,238.45,Consumo,Otros,NaN,Tarjeta de débito,NaN
2,110504306,Valentina Romero,Usuario con ingresos estacionales,2026-07-01,2026-07-31,2026-07-02,cambio de aceite,602.64,Consumo,Transporte,Esencial,Transferencia bancaria,NaN
3,110504306,Valentina Romero,Usuario con ingresos estacionales,2026-07-01,2026-07-31,2026-07-02,venta por internet,1935.92,Ingreso,NaN,NaN,Transferencia bancaria,NaN
4,110504306,Valentina Romero,Usuario con ingresos estacionales,2026-07-01,2026-07-31,2026-07-03,compré pescado,351.15,Consumo,Alimentación,Esencial,Efectivo,NaN


In [2]:
df.tail()

,user_id,nombre,arquetipo_perfil,periodo_inicio,periodo_fin,fecha,descripcion,monto,tipo_movimiento,categoria,grupo,tipo_pago,tasa_de_interes_de_la_tarjeta
4671,975413421,Andrea Flores,Usuario que pierde su empleo,2026-07-01,2026-07-31,2026-07-27,frutas y verduras,743.64,Consumo,Alimentación,Esencial,Tarjeta de crédito,41.9
4672,975413421,Andrea Flores,Usuario que pierde su empleo,2026-07-01,2026-07-31,2026-07-28,fui por el mandado,702.10,Consumo,Alimentación,Esencial,Tarjeta de débito,NaN
4673,975413421,Andrea Flores,Usuario que pierde su empleo,2026-07-01,2026-07-31,2026-07-28,pintura para la casa,1074.26,Consumo,Vivienda,Esencial,Tarjeta de débito,NaN
4674,975413421,Andrea Flores,Usuario que pierde su empleo,2026-07-01,2026-07-31,2026-07-29,supermercado,268.38,Consumo,Alimentación,Esencial,Transferencia bancaria,NaN
4675,975413421,Andrea Flores,Usuario que pierde su empleo,2026-07-01,2026-07-31,2026-07-29,compré sandalias,470.92,Consumo,Compras personales,Discrecional,Efectivo,NaN


### **Exploración de las columnas**

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4676 entries, 0 to 4675
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   user_id                        4676 non-null   int64  
 1   nombre                         4676 non-null   str    
 2   arquetipo_perfil               4676 non-null   str    
 3   periodo_inicio                 4676 non-null   str    
 4   periodo_fin                    4676 non-null   str    
 5   fecha                          4676 non-null   str    
 6   descripcion                    4676 non-null   str    
 7   monto                          4676 non-null   float64
 8   tipo_movimiento                4676 non-null   str    
 9   categoria                      4148 non-null   str    
 10  grupo                          4021 non-null   str    
 11  tipo_pago                      4676 non-null   str    
 12  tasa_de_interes_de_la_tarjeta  824 non-null    float64
dtyp

In [6]:
print("Número de filas:", df.shape[0])
print("Número de columnas:", df.shape[1])
print("Número de usuarios:", df["user_id"].nunique())

print("\nColumnas:")
print(df.columns.tolist())

Número de filas: 4676
Número de columnas: 13
Número de usuarios: 110

Columnas:
['user_id', 'nombre', 'arquetipo_perfil', 'periodo_inicio', 'periodo_fin', 'fecha', 'descripcion', 'monto', 'tipo_movimiento', 'categoria', 'grupo', 'tipo_pago', 'tasa_de_interes_de_la_tarjeta']


In [8]:
df.sample(10, random_state=42)

,user_id,nombre,arquetipo_perfil,periodo_inicio,periodo_fin,fecha,descripcion,monto,tipo_movimiento,categoria,grupo,tipo_pago,tasa_de_interes_de_la_tarjeta
2493,620531776,Emiliano García,Usuario con varias señales de estrés financiero simultáneas,2026-07-01,2026-07-31,2026-07-30,leche y huevos,427.33,Consumo,Alimentación,Esencial,Tarjeta de crédito,63.2
33,110504306,Valentina Romero,Usuario con ingresos estacionales,2026-07-01,2026-07-31,2026-07-24,seguro del coche,131.58,Consumo,Transporte,Esencial,Tarjeta de débito,NaN
3683,814760565,Eduardo López,Usuario con ingresos bajos pero excelente administración,2026-07-01,2026-07-31,2026-07-25,abono a la hipoteca,276.29,Pago de deuda,NaN,NaN,Transferencia bancaria,NaN
4129,864934246,Santiago Chávez,Usuario con ingresos variables,2026-07-01,2026-07-31,2026-07-14,cambio de aceite,229.33,Consumo,Transporte,Esencial,Tarjeta de débito,NaN
3193,732100362,Carolina Díaz,Usuario con ingresos altos pero sin ahorro,2026-07-01,2026-07-31,2026-07-25,camión,438.92,Consumo,Transporte,Esencial,Tarjeta de débito,NaN
4374,897310301,Emiliano Torres,Usuario que paga siempre el mínimo de su tarjeta,2026-07-01,2026-07-31,2026-07-15,pago de la quincena,3721.21,Ingreso,NaN,NaN,Transferencia bancaria,NaN
2716,681738546,Leonardo López,Usuario que gasta más de lo que gana,2026-07-01,2026-07-31,2026-07-28,perfume,756.21,Consumo,Compras personales,Discrecional,Efectivo,NaN
2303,601498537,Valentina García,Usuario que paga siempre el mínimo de su tarjeta,2026-07-01,2026-07-31,2026-07-14,pago de electricidad,594.97,Consumo,Vivienda,Esencial,Tarjeta de débito,NaN
3307,749041971,Javier Domínguez,Usuario que pierde su empleo,2026-07-01,2026-07-31,2026-07-07,verdura para la semana,308.45,Consumo,Alimentación,Esencial,Tarjeta de débito,NaN
2875,684892124,Javier Reyes,Usuario que invierte mientras utiliza crédito para gastos básicos,2026-07-01,2026-07-31,2026-07-31,invertí en CETES,1968.63,Ahorro e inversión,NaN,NaN,Transferencia bancaria,NaN


## **Limpieza de datos**

### **Exploración de valores nulos**

In [11]:
nulos = pd.DataFrame({
    "nulos": df.isna().sum(),
    "porcentaje": (df.isna().mean() * 100).round(2)
})

nulos

,nulos,porcentaje
user_id,0,0.00
nombre,0,0.00
periodo_inicio,0,0.00
periodo_fin,0,0.00
fecha,0,0.00
descripcion,0,0.00
monto,0,0.00
tipo_movimiento,0,0.00
categoria,528,11.29
grupo,655,14.01


In [12]:
df[["categoria", "grupo", "tasa_de_interes_de_la_tarjeta"]].isna().sum()

categoria                         528
grupo                             655
tasa_de_interes_de_la_tarjeta    3852
dtype: int64

##### **nan en categoria**

In [14]:
pd.crosstab(
    df["tipo_movimiento"],
    df["categoria"].isna(),
    margins=True
)

categoria,False,True,All
tipo_movimiento,,,
Ahorro e inversión,0,109,109
Consumo,4148,0,4148
Ingreso,0,255,255
Pago de deuda,0,164,164
All,4148,528,4676


In [15]:
df[
    (df["tipo_movimiento"] != "Consumo") &
    (df["categoria"].notna())
].shape[0]

0

Los nan se encuentran presentes en Ahorro e Inversión, Ingreso y Pago de deuda, las cuales son dimensiones que estructuralmente no tienen categroría porque no son consumo.

##### **nan en grupo**

In [17]:
pd.crosstab(
    df["tipo_movimiento"],
    df["grupo"].isna(),
    margins=True
)

grupo,False,True,All
tipo_movimiento,,,
Ahorro e inversión,0,109,109
Consumo,4021,127,4148
Ingreso,0,255,255
Pago de deuda,0,164,164
All,4021,655,4676


In [18]:
df[df["categoria"] == "Otros"]["grupo"].value_counts(dropna=False)

grupo
NaN    127
Name: count, dtype: int64

Hay un total de 655 nan en grupo. 528 son de las dimensiones diferentes a consumo que no necesitan "categoría" y por lo tanto no pueden clasificarse en "esencial" o "discrecional" + 127 pertenecientes a "otros", que por default no puedes categorizarse.

Por lo tanto, al no ser errores, no deben imputarse, puesto que tienen sentido lógico dentro de la estructura.


#### **Exploración tasa de interés**

Tada de interes unicamente sale cuando tipo_pago = tarjeta de credito

In [19]:
pd.crosstab(
    df["tipo_pago"],
    df["tasa_de_interes_de_la_tarjeta"].isna(),
    margins=True
)

tasa_de_interes_de_la_tarjeta,False,True,All
tipo_pago,,,
Efectivo,0,573,573
Tarjeta de crédito,824,0,824
Tarjeta de débito,0,1867,1867
Transferencia bancaria,0,1412,1412
All,824,3852,4676


In [20]:
credito_sin_tasa = df[
    (df["tipo_pago"] == "Tarjeta de crédito") &
    (df["tasa_de_interes_de_la_tarjeta"].isna())
]

tasa_sin_credito = df[
    (df["tipo_pago"] != "Tarjeta de crédito") &
    (df["tasa_de_interes_de_la_tarjeta"].notna())
]

print("Créditos sin tasa:", len(credito_sin_tasa))
print("Operaciones no crédito con tasa:", len(tasa_sin_credito))

Créditos sin tasa: 0
Operaciones no crédito con tasa: 0


No se debe imputar esta variable puesto que en todos los casos la tasa de interés sí aparece unicamente en entradas de tarjeta de credito como tipo de pago.

#### **Duplicados**

In [21]:
duplicados = df.duplicated().sum()

print("Registros duplicados:", duplicados)

Registros duplicados: 0


#### **Validación de identificadores del usuario**

In [22]:
print("Usuarios únicos:", df["user_id"].nunique())
print("IDs nulos:", df["user_id"].isna().sum())

Usuarios únicos: 110
IDs nulos: 0


In [25]:
validacion_usuario = (
    df.groupby("user_id")
      .agg(
          nombres=("nombre", "nunique"),
          perfiles=("arquetipo_perfil", "nunique")
      )
)

validacion_usuario.describe()

,nombres,perfiles
count,110.0,110.0
mean,1.0,1.0
std,0.0,0.0
min,1.0,1.0
25%,1.0,1.0
50%,1.0,1.0
75%,1.0,1.0
max,1.0,1.0


In [26]:
validacion_usuario[
    (validacion_usuario["nombres"] > 1) |
    (validacion_usuario["perfiles"] > 1)
]

,nombres,perfiles
user_id,,


#### **Validación de fechas**

In [32]:
columnas_fecha = [
    "periodo_inicio",
    "periodo_fin",
    "fecha"
]

for columna in columnas_fecha:
    df[columna] = pd.to_datetime(
        df[columna],
        format="%d-%m-%Y",
        errors="raise"
    )

In [34]:
for columna in columnas_fecha:
    print(
        columna,
        df[columna].min(),
        df[columna].max()
    )

periodo_inicio 2026-07-01 00:00:00 2026-07-01 00:00:00
periodo_fin 2026-07-31 00:00:00 2026-07-31 00:00:00
fecha 2026-07-01 00:00:00 2026-07-31 00:00:00


In [35]:
fuera_periodo = df[
    (df["fecha"] < df["periodo_inicio"]) |
    (df["fecha"] > df["periodo_fin"])
]

print("Transacciones fuera del periodo:", len(fuera_periodo))

Transacciones fuera del periodo: 0


#### **Validación de monto**

In [36]:
df["monto"] = pd.to_numeric(
    df["monto"],
    errors="raise"
).round(2)

In [37]:
print("Montos nulos:", df["monto"].isna().sum())
print("Montos en cero:", (df["monto"] == 0).sum())
print("Montos negativos:", (df["monto"] < 0).sum())

Montos nulos: 0
Montos en cero: 0
Montos negativos: 0


In [38]:
df["monto"].describe()

count      4676.000000
mean       1375.177825
std        6411.691000
min          20.000000
25%         208.932500
50%         425.475000
75%         852.087500
max      338891.250000
Name: monto, dtype: float64

#### **Outliers**

In [39]:
Q1 = df["monto"].quantile(0.25)
Q3 = df["monto"].quantile(0.75)

IQR = Q3 - Q1

limite_inferior = Q1 - 1.5 * IQR
limite_superior = Q3 + 1.5 * IQR

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("Límite inferior:", limite_inferior)
print("Límite superior:", limite_superior)

Q1: 208.9325
Q3: 852.0875
IQR: 643.155
Límite inferior: -755.8
Límite superior: 1816.82


In [40]:
outliers = df[
    (df["monto"] < limite_inferior) |
    (df["monto"] > limite_superior)
]

print("Número de posibles outliers:", len(outliers))

outliers.sort_values(
    "monto",
    ascending=False
).head(20)

Número de posibles outliers: 474


,user_id,nombre,arquetipo_perfil,periodo_inicio,periodo_fin,fecha,descripcion,monto,tipo_movimiento,categoria,grupo,tipo_pago,tasa_de_interes_de_la_tarjeta
2114,549756305,Leonardo Castillo,Usuario que recibe una herencia,2026-07-01,2026-07-31,2026-07-10,herencia recibida,338891.25,Ingreso,NaN,NaN,Transferencia bancaria,NaN
2790,684032941,Paola Jiménez,Usuario con ingresos altos pero sin ahorro,2026-07-01,2026-07-31,2026-07-15,nómina,86753.55,Ingreso,NaN,NaN,Transferencia bancaria,NaN
2822,684032941,Paola Jiménez,Usuario con ingresos altos pero sin ahorro,2026-07-01,2026-07-31,2026-07-30,me pagaron del trabajo,78332.36,Ingreso,NaN,NaN,Transferencia bancaria,NaN
1470,419239795,Paola Flores,Usuario que recibe una herencia,2026-07-01,2026-07-31,2026-07-25,herencia recibida,64554.49,Ingreso,NaN,NaN,Transferencia bancaria,NaN
2123,549756305,Leonardo Castillo,Usuario que recibe una herencia,2026-07-01,2026-07-31,2026-07-16,invertí en un pagaré,52357.48,Ahorro e inversión,NaN,NaN,Transferencia bancaria,NaN
2935,708223369,Fernanda Moreno,Usuario que recibe una herencia,2026-07-01,2026-07-31,2026-07-08,herencia recibida,49401.91,Ingreso,NaN,NaN,Transferencia bancaria,NaN
1291,394877417,Ricardo Moreno,Usuario que invierte mientras utiliza crédito para gastos básicos,2026-07-01,2026-07-31,2026-07-15,depósito de mi sueldo,46717.28,Ingreso,NaN,NaN,Transferencia bancaria,NaN
4583,931697199,Valentina Martínez,Usuario con ingresos estacionales,2026-07-01,2026-07-31,2026-07-18,venta por internet,44769.49,Ingreso,NaN,NaN,Efectivo,NaN
1330,394877417,Ricardo Moreno,Usuario que invierte mientras utiliza crédito para gastos básicos,2026-07-01,2026-07-31,2026-07-30,me cayó la nómina,44211.13,Ingreso,NaN,NaN,Transferencia bancaria,NaN
1114,360278346,Renata Romero,Usuario con deuda creciente pese a ingresos estables,2026-07-01,2026-07-31,2026-07-30,sueldo,41309.01,Ingreso,NaN,NaN,Transferencia bancaria,NaN


El ingreso de $300,000 puede ser extraño respecto a la distribución general pero perfectamente válido para alguno de los perfiles.

In [42]:
#Inspecciona por tipo de movimientos
for tipo, datos in df.groupby("tipo_movimiento"):

    Q1 = datos["monto"].quantile(0.25)
    Q3 = datos["monto"].quantile(0.75)
    IQR = Q3 - Q1

    limite_superior = Q3 + 1.5 * IQR

    n_outliers = (datos["monto"] > limite_superior).sum()

    print(
        f"{tipo}: "
        f"{n_outliers} posibles outliers | "
        f"Límite: ${limite_superior:,.2f} | "
        f"Máximo: ${datos['monto'].max():,.2f}"
    )

Ahorro e inversión: 9 posibles outliers | Límite: $4,448.41 | Máximo: $52,357.48
Consumo: 241 posibles outliers | Límite: $1,427.89 | Máximo: $6,124.07
Ingreso: 14 posibles outliers | Límite: $34,762.49 | Máximo: $338,891.25
Pago de deuda: 11 posibles outliers | Límite: $5,075.07 | Máximo: $15,535.59


#### **Tratamiento de las variables de texto**

In [45]:
columnas_texto = [
    "nombre",
    "arquetipo_perfil",
    "descripcion",
    "tipo_movimiento",
    "categoria",
    "grupo",
    "tipo_pago"
]

for columna in columnas_texto:

    mask = df[columna].notna()

    df.loc[mask, columna] = (
        df.loc[mask, columna]
          .astype(str)
          .str.strip()
          .str.replace(r"\s+", " ", regex=True)
    )

#### **Creación variable descripcion_limpia**

In [46]:
def limpiar_descripcion(texto):

    texto = unicodedata.normalize("NFC", texto)

    texto = texto.strip().lower()

    texto = " ".join(texto.split())

    return texto


df["descripcion_limpia"] = df["descripcion"].apply(
    limpiar_descripcion
)

In [51]:
df[
    ["descripcion", "descripcion_limpia"]
].sample(10, random_state=30)

,descripcion,descripcion_limpia
2489,seguro de gastos médicos,seguro de gastos médicos
1625,fruta para la casa,fruta para la casa
3047,metrobús,metrobús
3911,ingreso por un proyecto,ingreso por un proyecto
2844,pensión del coche,pensión del coche
270,metí dinero a Actinver,metí dinero a actinver
326,fui con el especialista,fui con el especialista
3636,una chamarra,una chamarra
1140,Dropbox,dropbox
1610,refrescos para la casa,refrescos para la casa


#### **Distribución de tipos de movimiento**

In [52]:
conteo_movimientos = (
    df["tipo_movimiento"]
    .value_counts()
    .to_frame("cantidad")
)

conteo_movimientos["porcentaje"] = (
    conteo_movimientos["cantidad"]
    / len(df)
    * 100
).round(2)

conteo_movimientos

,cantidad,porcentaje
tipo_movimiento,,
Consumo,4148,88.71
Ingreso,255,5.45
Pago de deuda,164,3.51
Ahorro e inversión,109,2.33


Aunque hay un fuerte desbalance de clases, esto es normal pues en la vida real, las personas suelen tener muchos más gastos que ingresos. Es decir la cantidad total invertida en gastos/consumo se divide en más partes que la cantidad total del ingreso.

#### **Distribución de categorías de consumo**

In [53]:
df_consumo = df[
    df["tipo_movimiento"] == "Consumo"
].copy()

In [54]:
distribucion_categorias = (
    df_consumo["categoria"]
    .value_counts()
    .to_frame("cantidad")
)

distribucion_categorias["porcentaje"] = (
    distribucion_categorias["cantidad"]
    / len(df_consumo)
    * 100
).round(2)

distribucion_categorias

,cantidad,porcentaje
categoria,,
Alimentación,1609,38.79
Transporte,1064,25.65
Compras personales,327,7.88
Vivienda,261,6.29
Entretenimiento y ocio,224,5.40
Suscripciones digitales,167,4.03
Salud,144,3.47
Educación,140,3.38
Otros,127,3.06


#### **Distribución de grupos**

In [55]:
df_consumo["grupo"].value_counts(
    dropna=False
)

grupo
Esencial        3218
Discrecional     803
NaN              127
Name: count, dtype: int64

In [56]:
df_consumo["grupo"].value_counts(
    normalize=True,
    dropna=False
).mul(100).round(2)

grupo
Esencial        77.58
Discrecional    19.36
NaN              3.06
Name: proportion, dtype: float64

#### **Análisis de montos por movimiento**

In [57]:
df.groupby("tipo_movimiento")["monto"].agg([
    "count",
    "mean",
    "median",
    "std",
    "min",
    "max"
]).round(2)

,count,mean,median,std,min,max
tipo_movimiento,,,,,,
Ahorro e inversión,109,2756.38,1440.66,6245.35,150.83,52357.48
Consumo,4148,534.15,378.54,554.03,20.00,6124.07
Ingreso,255,14079.00,9794.24,23609.02,151.05,338891.25
Pago de deuda,164,1976.10,1281.20,2176.29,101.25,15535.59


#### **Montos por categoría**

In [58]:
df_consumo.groupby("categoria")["monto"].agg([
    "count",
    "mean",
    "median",
    "min",
    "max"
]).round(2).sort_values(
    "median",
    ascending=False
)

,count,mean,median,min,max
categoria,,,,,
Vivienda,261,1362.93,1171.63,20.00,5877.28
Compras personales,327,796.06,637.10,20.00,5167.83
Educación,140,668.07,550.28,33.06,2938.83
Salud,144,522.25,468.97,20.00,1584.36
Otros,127,452.75,389.06,20.00,2494.20
Alimentación,1609,502.61,377.18,20.00,6124.07
Entretenimiento y ocio,224,458.12,369.22,20.00,3516.10
Transporte,1064,375.17,286.18,20.00,3474.53
Viajes y vacaciones,85,309.30,195.00,35.83,1995.80


frecuencia de la categoría ≠ cantidad de dinero gastado

#### **Distribución por medio de pago**

In [60]:
df["tipo_pago"].unique()

<StringArray>
['Efectivo', 'Tarjeta de débito', 'Transferencia bancaria',
 'Tarjeta de crédito']
Length: 4, dtype: str

In [59]:
distribucion_pago = (
    df["tipo_pago"]
    .value_counts()
    .to_frame("cantidad")
)

distribucion_pago["porcentaje"] = (
    distribucion_pago["cantidad"]
    / len(df)
    * 100
).round(2)

distribucion_pago

,cantidad,porcentaje
tipo_pago,,
Tarjeta de débito,1867,39.93
Transferencia bancaria,1412,30.20
Tarjeta de crédito,824,17.62
Efectivo,573,12.25


#### **Análisis de tipo de pago = tarjeta de crédito**

In [61]:
df_credito = df[
    df["tipo_pago"] == "Tarjeta de crédito"
].copy()

In [63]:
df_credito["tasa_de_interes_de_la_tarjeta"].describe().round(2)

count    824.00
mean      39.93
std       12.10
min       14.70
25%       29.70
50%       41.35
75%       49.00
max       63.20
Name: tasa_de_interes_de_la_tarjeta, dtype: float64

#### **Diversidad en las descripciones**

In [64]:
print(
    "Transacciones:",
    len(df)
)

print(
    "Descripciones únicas:",
    df["descripcion"].nunique()
)

print(
    "Descripciones limpias únicas:",
    df["descripcion_limpia"].nunique()
)

Transacciones: 4676
Descripciones únicas: 535
Descripciones limpias únicas: 535


In [65]:
#comprueba la repitición
frecuencia_descripciones = (
    df["descripcion_limpia"]
    .value_counts()
    .reset_index()
)

frecuencia_descripciones.columns = [
    "descripcion_limpia",
    "frecuencia"
]

frecuencia_descripciones.head(20)

,descripcion_limpia,frecuencia
0,agua embotellada,75
1,botanas para la casa,73
2,jugos y agua,69
3,refrescos para la casa,67
4,snacks del súper,65
5,verdura para la semana,61
6,fruta para la casa,60
7,pescado,59
8,frutas y verduras,57
9,compré verduras,54


#### **Contradicciones semánticas**

Comprueba si una descripción tiene diferentes etiquetas

In [67]:
#para tipo de movimiento
conflictos_movimiento = (
    df.groupby("descripcion_limpia")
      ["tipo_movimiento"]
      .nunique()
)

conflictos_movimiento = conflictos_movimiento[
    conflictos_movimiento > 1
]

print(
    "Descripciones con múltiples tipos de movimiento:",
    len(conflictos_movimiento)
)

Descripciones con múltiples tipos de movimiento: 0


In [68]:
#para las categorías
conflictos_categoria = (
    df_consumo.groupby("descripcion_limpia")
              ["categoria"]
              .nunique()
)

conflictos_categoria = conflictos_categoria[
    conflictos_categoria > 1
]

print(
    "Descripciones con múltiples categorías:",
    len(conflictos_categoria)
)

Descripciones con múltiples categorías: 0


#### **Validación final**

In [69]:
print("Dimensiones finales:", df.shape)
print("Usuarios:", df["user_id"].nunique())
print("Duplicados:", df.duplicated().sum())
print("Descripciones vacías:", (df["descripcion_limpia"] == "").sum())
print("Montos <= 0:", (df["monto"] <= 0).sum())

print("\nNulos:")
print(df.isna().sum())

Dimensiones finales: (4676, 14)
Usuarios: 110
Duplicados: 0
Descripciones vacías: 0
Montos <= 0: 0

Nulos:
user_id                             0
nombre                              0
arquetipo_perfil                    0
periodo_inicio                      0
periodo_fin                         0
fecha                               0
descripcion                         0
monto                               0
tipo_movimiento                     0
categoria                         528
grupo                             655
tipo_pago                           0
tasa_de_interes_de_la_tarjeta    3852
descripcion_limpia                  0
dtype: int64


### **Exportación del dataset limpio**

In [70]:
df.to_csv(
    "dataset_finance_ai_110_usuarios_limpio.csv",
    index=False,
    encoding="utf-8"
)

## **Elaboración del Clasificador**

In [40]:
df_clasificacion = df[
    df["tipo_movimiento"] == "Consumo"
].copy()

In [41]:
# ============================================
# Categorías oficiales de consumo del sistema
# ============================================

CATEGORIAS_CONSUMO = [
    "Vivienda",
    "Alimentación",
    "Transporte",
    "Salud",
    "Educación",
    "Entretenimiento y ocio",
    "Suscripciones digitales",
    "Compras personales",
    "Viajes y vacaciones",
    "Otros"
]

# Implementación de motor financiero

## 1.- Recepción de archivo JSON de prueba ya clasificado manualmente 

In [42]:
import json
import random
from pathlib import Path

def cargar_datos_clasificados(carpeta="datos"):
    """
    Selecciona aleatoriamente uno de los archivos JSON clasificados,
    lo convierte a un diccionario y lo devuelve.
    """

    carpeta_json = Path("perfiles_financieros_pruebas_clasificados_para_motor_financiero")

    numero = random.randint(1, 15)

    nombre_archivo = f"{numero:02d}_clasificado_manualmente.json"

    ruta_archivo = carpeta_json / nombre_archivo

    with open(ruta_archivo, "r", encoding="utf-8") as archivo:
        datos = json.load(archivo)

    print(f"Archivo seleccionado: {nombre_archivo}")

    return datos


datos_clasificados = cargar_datos_clasificados()
print (datos_clasificados)

Archivo seleccionado: 10_clasificado_manualmente.json
{'usuario': {'nombre': 'Iván Rosales'}, 'periodo': {'inicio': '2026-07-01', 'fin': '2026-07-31'}, 'ingresos': [{'fecha': '2026-07-01', 'descripcion': 'Salario', 'monto': 17000}], 'transacciones': [{'fecha': '2026-07-01', 'descripcion': 'Renta departamento', 'monto': 7500, 'forma_pago': 'Transferencia bancaria', 'clasificacion': ['Consumo'], 'categoria': 'Vivienda', 'grupo': 'Esencial'}, {'fecha': '2026-07-03', 'descripcion': 'Pago tarjeta BBVA', 'monto': 4500, 'forma_pago': 'Transferencia bancaria', 'clasificacion': ['Pago de deuda']}, {'fecha': '2026-07-05', 'descripcion': 'Pago préstamo personal', 'monto': 3800, 'forma_pago': 'Transferencia bancaria', 'clasificacion': ['Pago de deuda']}, {'fecha': '2026-07-07', 'descripcion': 'Despensa Soriana', 'monto': 2200, 'forma_pago': 'Tarjeta de crédito', 'tasa_de_interes_de_la_tarjeta': 92, 'clasificacion': ['Consumo', 'Pago de deuda'], 'categoria': 'Alimentación', 'grupo': 'Esencial'}, {'

## 2.- Generación de varibles globales

In [43]:
def calcular_ingreso_mensual(datos_clasificados):
    """
    Calcula el ingreso mensual del usuario.

    Parámetros:
        datos_clasificados (dict): Diccionario con la información del usuario.

    Retorna:
        float: Suma de todos los ingresos registrados durante el periodo.
    """

    ingreso_mensual = 0.0

    for ingreso in datos_clasificados["ingresos"]:
        ingreso_mensual += ingreso["monto"]

    return ingreso_mensual

ingreso_mensual = calcular_ingreso_mensual(datos_clasificados)

print(f"Ingreso mensual: ${ingreso_mensual:,.2f}")

Ingreso mensual: $17,000.00


In [44]:
def calcular_consumo_total_mensual(datos_clasificados):
    """
    Calcula el consumo total mensual del usuario.

    Se consideran únicamente las transacciones cuya clasificación
    contenga exclusivamente la etiqueta "Consumo".

    Parámetros:
        datos_clasificados (dict): Diccionario con las transacciones clasificadas.

    Retorna:
        float: Consumo total mensual.
    """

    consumo_total_mensual = 0.0

    for transaccion in datos_clasificados["transacciones"]:

        clasificacion = transaccion["clasificacion"]

        if (
            "Consumo" in clasificacion
            and len(clasificacion) == 1
        ):
            consumo_total_mensual += transaccion["monto"]

    return consumo_total_mensual

consumo_total_mensual = calcular_consumo_total_mensual(datos_clasificados)

print(f"Consumo total mensual: ${consumo_total_mensual:,.2f}")

Consumo total mensual: $7,500.00


In [45]:
def calcular_pago_mensual_de_deudas(datos_clasificados):
    """
    Calcula el pago mensual de deudas del usuario.

    Se consideran todas las transacciones cuya clasificación
    incluya la etiqueta "Pago de deuda", independientemente
    de si también pertenecen a otra clasificación.

    Parámetros:
        datos_clasificados (dict): Diccionario con las transacciones clasificadas.

    Retorna:
        float: Pago mensual de deudas.
    """

    pago_mensual_de_deudas = 0.0

    for transaccion in datos_clasificados["transacciones"]:

        clasificacion = transaccion["clasificacion"]

        if "Pago de deuda" in clasificacion:
            pago_mensual_de_deudas += transaccion["monto"]

    return pago_mensual_de_deudas

pago_mensual_de_deudas = calcular_pago_mensual_de_deudas(datos_clasificados)

print(f"Pago mensual de deudas: ${pago_mensual_de_deudas:,.2f}")

Pago mensual de deudas: $24,530.00


In [46]:
def calcular_ahorro_e_inversion_total(datos_clasificados):
    """
    Calcula el ahorro e inversión total del usuario.

    Se consideran todas las transacciones cuya clasificación
    incluya la etiqueta "Ahorro e inversión".

    Parámetros:
        datos_clasificados (dict): Diccionario con las transacciones clasificadas.

    Retorna:
        float: Ahorro e inversión total.
    """

    ahorro_e_inversion_total = 0.0

    for transaccion in datos_clasificados["transacciones"]:

        clasificacion = transaccion["clasificacion"]

        if "Ahorro e inversión" in clasificacion:
            ahorro_e_inversion_total += transaccion["monto"]

    return ahorro_e_inversion_total

ahorro_e_inversion_total = calcular_ahorro_e_inversion_total(datos_clasificados)

print(f"Ahorro e inversión total: ${ahorro_e_inversion_total:,.2f}")

Ahorro e inversión total: $0.00


In [47]:
def calcular_egreso_total(consumo_total_mensual, pago_mensual_de_deudas):
    """
    Calcula el egreso total del usuario.

    El egreso total corresponde a la suma del consumo total mensual
    y el pago mensual de deudas.

    Parámetros:
        consumo_total_mensual (float): Consumo total mensual del usuario.
        pago_mensual_de_deudas (float): Pago mensual de deudas del usuario.

    Retorna:
        float: Egreso total del usuario.
    """

    egreso_total = consumo_total_mensual + pago_mensual_de_deudas

    return egreso_total

egreso_total = calcular_egreso_total(
    consumo_total_mensual,
    pago_mensual_de_deudas
)

print(f"Egreso total: ${egreso_total:,.2f}")

Egreso total: $32,030.00


In [48]:
def calcular_balance_mensual(ingreso_mensual, egreso_total):
    """
    Calcula el balance mensual del usuario.

    El balance mensual corresponde a la diferencia entre el ingreso
    mensual y el egreso total del periodo.

    Parámetros:
        ingreso_mensual (float): Ingreso mensual del usuario.
        egreso_total (float): Egreso total del usuario.

    Retorna:
        float: Balance mensual del usuario.
    """

    balance_mensual = ingreso_mensual - egreso_total

    return balance_mensual

balance_mensual = calcular_balance_mensual(
    ingreso_mensual,
    egreso_total
)

print(f"Balance mensual: ${balance_mensual:,.2f}")

Balance mensual: $-15,030.00


In [49]:
def calcular_consumo_total_por_categoria(datos_clasificados):
    """
    Calcula el consumo total por categoría del usuario.

    Se consideran todas las transacciones cuya clasificación
    incluya la etiqueta "Consumo", independientemente de que
    también estén clasificadas como "Pago de deuda".

    Parámetros:
        datos_clasificados (dict): Diccionario con las transacciones clasificadas.

    Retorna:
        float: Consumo total por categoría.
    """

    consumo_total_por_categoria = 0.0

    for transaccion in datos_clasificados["transacciones"]:

        clasificacion = transaccion["clasificacion"]

        if "Consumo" in clasificacion:
            consumo_total_por_categoria += transaccion["monto"]

    return consumo_total_por_categoria

consumo_total_por_categoria = calcular_consumo_total_por_categoria(datos_clasificados)

print(f"Consumo total por categoría: ${consumo_total_por_categoria:,.2f}")

Consumo total por categoría: $19,530.00


In [50]:
def calcular_gasto_por_categoria(datos_clasificados):
    """
    Calcula el gasto acumulado por categoría de consumo.

    Se consideran todas las transacciones cuya clasificación
    incluya la etiqueta "Consumo", independientemente de que
    también estén clasificadas como "Pago de deuda".

    Parámetros:
        datos_clasificados (dict): Diccionario con las transacciones clasificadas.

    Retorna:
        dict: Diccionario con el gasto acumulado por cada categoría.
    """

    # Inicializar todas las categorías en 0.0
    gasto_por_categoria = {
        categoria: 0.0
        for categoria in CATEGORIAS_CONSUMO
    }

    # Recorrer las transacciones
    for transaccion in datos_clasificados["transacciones"]:

        clasificacion = transaccion["clasificacion"]

        if "Consumo" in clasificacion:

            categoria = transaccion["categoria"]

            gasto_por_categoria[categoria] += transaccion["monto"]

    return gasto_por_categoria

gasto_por_categoria = calcular_gasto_por_categoria(datos_clasificados)

print("Gasto por categoría:")

for categoria, monto in gasto_por_categoria.items():
    print(f"{categoria}: ${monto:,.2f}")


Gasto por categoría:
Vivienda: $8,280.00
Alimentación: $2,200.00
Transporte: $1,600.00
Salud: $900.00
Educación: $0.00
Entretenimiento y ocio: $2,450.00
Suscripciones digitales: $0.00
Compras personales: $4,100.00
Viajes y vacaciones: $0.00
Otros: $0.00


In [51]:
def calcular_distribucion_porcentual_gasto(
    gasto_por_categoria,
    consumo_total_por_categoria
):
    """
    Calcula la distribución porcentual del gasto por categoría.

    La distribución porcentual representa el porcentaje que ocupa
    cada categoría respecto al consumo total por categoría.

    Parámetros:
        gasto_por_categoria (dict): Gasto acumulado por categoría.
        consumo_total_por_categoria (float): Consumo total por categoría.

    Retorna:
        dict: Distribución porcentual del gasto por categoría.
    """

    # Inicializar todas las categorías en 0.0
    distribucion_porcentual_gasto = {
        categoria: 0.0
        for categoria in CATEGORIAS_CONSUMO
    }

    # Evitar división entre cero
    if consumo_total_por_categoria == 0.0:
        return distribucion_porcentual_gasto

    # Calcular la distribución porcentual
    for categoria in CATEGORIAS_CONSUMO:

        distribucion_porcentual_gasto[categoria] = (
            gasto_por_categoria[categoria]
            / consumo_total_por_categoria
        ) * 100

    return distribucion_porcentual_gasto

distribucion_porcentual_gasto = calcular_distribucion_porcentual_gasto(
    gasto_por_categoria,
    consumo_total_por_categoria
)

print("Distribución porcentual del gasto:")

for categoria, porcentaje in distribucion_porcentual_gasto.items():
    print(f"{categoria}: {porcentaje:.2f}%")

Distribución porcentual del gasto:
Vivienda: 42.40%
Alimentación: 11.26%
Transporte: 8.19%
Salud: 4.61%
Educación: 0.00%
Entretenimiento y ocio: 12.54%
Suscripciones digitales: 0.00%
Compras personales: 20.99%
Viajes y vacaciones: 0.00%
Otros: 0.00%


In [52]:
def calcular_tasa_interes_promedio_ponderada(datos_clasificados):
    """
    Calcula la tasa de interés promedio ponderada de las compras
    realizadas con tarjeta de crédito.

    El promedio se calcula utilizando el monto de cada transacción
    como peso.

    Parámetros:
        datos_clasificados (dict): Diccionario con las transacciones clasificadas.

    Retorna:
        float: Tasa de interés promedio ponderada.
    """

    monto_total_credito = 0.0
    suma_ponderada = 0.0

    # Recorrer las transacciones
    for transaccion in datos_clasificados["transacciones"]:

        if transaccion["forma_pago"] == "Tarjeta de crédito":

            monto = transaccion["monto"]
            tasa = transaccion["tasa_de_interes_de_la_tarjeta"]

            monto_total_credito += monto
            suma_ponderada += monto * tasa

    # Evitar división entre cero
    if monto_total_credito == 0.0:
        return 0.0

    tasa_interes_promedio_ponderada = (
        suma_ponderada / monto_total_credito
    )

    return tasa_interes_promedio_ponderada

tasa_interes_promedio_ponderada = calcular_tasa_interes_promedio_ponderada(
    datos_clasificados
)

print(
    f"Tasa de interés promedio ponderada: "
    f"{tasa_interes_promedio_ponderada:.2f}%"
)

Tasa de interés promedio ponderada: 92.00%
